### PR TITLE
Add OpenID Connect as authentication

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

**Comment by Hye-Dev** [comment]
I tested this and there appears to be a bug. If you authenticate with OIDC, but then try to access resources on a project other than 'default', it states that the request is not authorized (doing lxc list --project=projectname for example).

**Comment by monstermunchkin** [comment]
> I tested this and there appears to be a bug. If you authenticate with OIDC, but then try to access resources on a project other than 'default', it states that the request is not authorized (doing lxc list --project=projectname for example).

Thanks for pointing that out. It was an issue in the client, and has been fixed now.

**Comment by stgraber** [comment]
Okay, so in general I'm happy with the configuration keys and overall functionality of this.
I'd however like to try to isolate things a bit more so it's easier to follow, that's why I started with the introduction of the candid and oidc packages on the server side.

On that front, I think most of the request handling in lxd/daemon.go (the chunk starting with the header check) can be moved over to the package and be called through a function on the verifier struct, similar to the Auth function on the candidVerifier.

The other thing that's been bothering me but that I haven't had time to sort out yet, is the need to extend our errors to include metadata. That seems a bit odd. The addition of X-LXD-oidc also feels wrong to me.

Instead I think we should be able to rely on our existing X-LXD-authenticated, which effectively tells LXD to require authentication for every API endpoint, even those that would be accessible unauthenticated. Then we'd hit `/1.0` which will fail with 403 but on which we can then set some additional headers, namely:
 - X-LXD-OIDC-issuer
 - X-LXD-OIDC-clientid
 - X-LXD-OIDC-audience

This then gives us the information we need to setup the OIDC client and then try again.

That should let us drop that response change from the branch and also drop the extra client-side header (X-LXD-oidc) in favor of just server-side headers being emitted. Which should also let us drop the `shared/oidc.go` as part of that which was something else that was bothering me with this branch :)

If you can get that done, we should end up with a branch that's very self-contained on the server side and somewhat self-contained on the client side. I'll still want to take a deeper look at the logic in the client and CLI to see if I'm happy with the structure, but we're getting there.

File: lxd/auth/candid/candid.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
+package candid
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/lxc/lxd/lxd/response"
```
Comment (tomponline) [inline lxd/auth/candid/candid.go:9]: This should come after the external deps.

**Comment by monstermunchkin** [comment]
@stgraber good to go

**Review by stgraber — APPROVED** [review]
Have done a few more edits to re-group code a bit better and fix a few tiny issues here and there, I'm now happy with this. Once it's green enough, we can merge.

**Pull request merged**